FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1331 @@
+          if (ReducedPrecisionMode) {
+            Value = IREmit->_Vector_FToI(8, 8, st0, Round_Host);
+          } else {
+            Value = IREmit->_F80Round(st0);
+          }
+          StoreStackValueAtTop_Slow(Value, false);
+        } else {
+          StackData.setTop(StackMemberInfo {.SourceDataSize = StackMember.SourceDataSize,
+                                            .SourceDataNode = nullptr,
+                                            .StackDataNode = ReducedPrecisionMode ? IREmit->_Vector_FToI(8, 8, StackMember.StackDataNode, Round_Host) :
+                                                                                    IREmit->_F80Round(StackMember.StackDataNode),
+                                            .InterpretAsFloat = false});
+        }
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80VBSLSTACK: {
+        // FIXME: I am sort of surprised that this is both used in reduced and normal precision modes.
+        const auto* Op = IROp->C<IROp_F80VBSLStack>();
+
+        // Multiplies two elements in the stack by offset.
+        auto* VecCond = CurrentIR.GetNode(Op->VectorMask);
+        auto StackOffset1 = Op->SrcStack1;
+        auto StackOffset2 = Op->SrcStack2;
+
+        const auto& [Valid1, StackMember1] = StackData.top(StackOffset1);
+        const auto& [Valid2, StackMember2] = StackData.top(StackOffset2);
+
+        MigrateToSlowPathIf(Valid1 != StackSlot::VALID || Valid2 != StackSlot::VALID);
+
+        if (SlowPath) {
+          auto* Value1 = LoadStackValueAtOffset_Slow(StackOffset1);
+          auto* Value2 = LoadStackValueAtOffset_Slow(StackOffset2);
+
+          StoreStackValueAtTop_Slow(IREmit->_VBSL(16, VecCond, Value1, Value2), StackOffset1 != 0 && StackOffset2 != 0);
+        } else {
+          StackData.setTop(StackMemberInfo {.SourceDataSize = StackMember1.SourceDataSize,
+                                            .SourceDataNode = nullptr,
+                                            .StackDataNode = IREmit->_VBSL(16, VecCond, StackMember1.StackDataNode, StackMember2.StackDataNode),
+                                            .InterpretAsFloat = false});
+        }
+
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      default: break;
+      }
+    }
+
+    // We need to write the registers before any branch in the block,
+    // so loop until a branch instruction is found. Add instructions _before_
+    // the branch instruction.
+    [[maybe_unused]] bool ExitFound = false; // used for assertion
+    for (auto [CodeNode, IROp] : CurrentIR.GetCode(BlockNode)) {
+      if (IsBlockExit(IROp->Op)) {
+        // Set write cursor to previous instruction
+        IREmit->SetWriteCursor(IREmit->UnwrapNode(CodeNode->Header.Previous));
+        ExitFound = true;
+        break;
+      }
+    }
+    LOGMAN_THROW_A_FMT(ExitFound, "No block exit found is bad!");
+    SynchronizeStackValues();
+    IREmit->SetWriteCursor(OriginalWriteCursor);
```

Comment:
drop this, no need to save/restore cursor

Reply:
Perfect!